README.md
Gedune
======

The Gedune, a modification for starfarer

ChangeLog.txt
Version 0.8 
	* .ship tweaks, improved boundries and weapon mounts.
	* Gave the gedune their very own ship and fighter engine styles. 
	* big nerf to flare PD. Those Duri were murderously scary vs frigates. Still are.
	* Adjusted colour of the Nuclear Burn system to match the new engine styles.
	* Added repeater with graphics!
	* Tenzen Tweaks
	* Gedune now <3 the flarePD
	* Spammy repeater added, fun gun.
	* Dahaki super shield nerf, FP boost, and phase repeater added, EMP gun. Awesome on a TP ship. Really happy with that one.
	* Tenzen now use speed burst system
	* Chua adjustments
	* Station now has stuff!
	* Added Byakuri class carrier, this ship will need balancing.
	* Changed Byakuri sprite and added new (phase)repeater sprite.
Version 0.9
	* Tweaked the Tychrea - it is now a fully fledged ship, still a secret though.
	* added two new built in weapons, the Explosive AEB and the Kinetic REB.
	*Added a new ship: The Kyirus interceptor frigate - armed with both new weapons and a nuclear burndrive.
Version 0.95
	* Added descriptions to everything excluding the Kyra drone.
	* rebalanced the nuclear burst weapon (formerly gedune plasma).
	* Further tweaking of the NC. lower damage and a longer charging rate.
Version 1
	* Re-made the Tychrea frigate. Is now sightly bigger, less weaponry and uses a drone system
	* added the wraithii drone, along with the wraith weapon - a high speed low yield blaster.

data/scripts/world/GeduneSpawnPoint.java
package data.scripts.world;

import com.fs.starfarer.api.campaign.*;
import data.scripts.world.BaseSpawnPoint;

public class GeduneSpawnPoint extends BaseSpawnPoint {

	public GeduneSpawnPoint(SectorAPI sector, LocationAPI location, float daysInterval,
			int maxFleets, SectorEntityToken anchor) {
		super(sector, location, daysInterval, maxFleets, anchor);
	}
   
	@Override
	protected CampaignFleetAPI spawnFleet() {
		float r = (float) Math.random();
		if (       (r -= .5f) < 0) {
			return raidFleet("scout");
		} else if ((r -= .4f) < 0) {
			return raidFleet("longRangeScout");
		} else if ((r -= .3f) < 0) {
			return raidFleet("raiders");
		} else if ((r -= .2f) < 0) {
			return Math.random() > 0.3 ? raidFleet("libers"): defendFleet("libers");
		} else if ((r -= .2f) < 0) {
         return l33tFleet(Math.random() > 0.1 ? raidFleet("elite"): defendFleet("elite"));
		} else {
			return Math.random() > 0.1 ? raidFleet("siege"): defendFleet("siege");
		}
	}

	private CampaignFleetAPI raidFleet(String fleetID) {
		CampaignFleetAPI fleet = getSector().createFleet("gedune", fleetID);
		getLocation().spawnFleet(getAnchor(), 0, 0, fleet);
		fleet.setPreferredResupplyLocation(getAnchor());
		fleet.addAssignment(FleetAssignment.RAID_SYSTEM, null, 10);
		fleet.addAssignment(FleetAssignment.DELIVER_RESOURCES, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DELIVER_SUPPLIES, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DELIVER_FUEL, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.GO_TO_LOCATION_AND_DESPAWN, getAnchor(), 1000);
		return fleet;
	}

	private CampaignFleetAPI defendFleet(String fleetID) {
		CampaignFleetAPI fleet = getSector().createFleet("gedune", fleetID);
		getLocation().spawnFleet(getAnchor(), 0, 0, fleet);
		fleet.setPreferredResupplyLocation(getAnchor());
		fleet.addAssignment(FleetAssignment.DEFEND_LOCATION, getAnchor(), 15);
		fleet.addAssignment(FleetAssignment.DELIVER_RESOURCES, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DELIVER_SUPPLIES, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DELIVER_FUEL, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DEFEND_LOCATION, getAnchor(), 15);
		fleet.addAssignment(FleetAssignment.DELIVER_RESOURCES, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DELIVER_SUPPLIES, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.DELIVER_FUEL, getAnchor(), 100);
		fleet.addAssignment(FleetAssignment.GO_TO_LOCATION_AND_DESPAWN, getAnchor(), 1000);
		return fleet;
	}

	private CampaignFleetAPI l33tFleet(CampaignFleetAPI fleet) {
		CargoAPI fcargo = fleet.getCargo();
		int t = fcargo.getTotalCrew();
		double ratio = Math.random() * 0.2 + 0.1;
		fcargo.removeCrew(CargoAPI.CrewXPLevel.REGULAR, t);
		fcargo.addCrew(CargoAPI.CrewXPLevel.ELITE, (int) (t * ratio));
		fcargo.addCrew(CargoAPI.CrewXPLevel.VETERAN, (int) (t * (0.25 - ratio / 2)));
		fcargo.addCrew(CargoAPI.CrewXPLevel.REGULAR, (int) (t * (0.75 - ratio / 2)));
		return fleet;
	}
}

data/scripts/world/GeduneStationManager.java
package data.scripts.world;

import java.awt.Color;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.CargoAPI;
import com.fs.starfarer.api.campaign.CargoStackAPI;
import com.fs.starfarer.api.campaign.FleetDataAPI;
import com.fs.starfarer.api.campaign.LocationAPI;
import com.fs.starfarer.api.campaign.SectorAPI;
import com.fs.starfarer.api.campaign.SectorEntityToken;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.fleet.FleetMemberType;

public class GeduneStationManager extends BaseSpawnPoint {
	private static final float FUEL_LIMIT = 1000.0f;
	private static final float SUPL_LIMIT = 800.0f; 
	private static final float FUEL_GEN = 5f; 
	private static boolean NAG = false; //debug messages
	
	//bonus Fuel is a balancing force. Meant to simulate the gedune
	//setting up temporary reactors in response to resource surplus. 
	private float bonusFuel = 0;
	
	private SectorEntityToken refactory;
	private CargoAPI refCargo;

	public GeduneStationManager(SectorAPI sector, LocationAPI location, float daysInterval
			        , SectorEntityToken anchor) {
		super(sector, location, daysInterval, 1, anchor);
		refactory = anchor;
		refCargo = refactory.getCargo();
	}

	@Override
	protected CampaignFleetAPI spawnFleet() {

		// bonus fuel is added gradually, 30 every turn
		refCargo.addFuel(FUEL_GEN + (bonusFuel<30 ? bonusFuel : 30));
		if(bonusFuel < 30) bonusFuel = 0;
		else			   bonusFuel -= 30;
		
		// if we have too little fuel and too much supplies, break supplies down for fuel
		if (refCargo.getFuel() < FUEL_LIMIT && refCargo.getSupplies() > SUPL_LIMIT){
			bonusFuel += 300;
			refCargo.removeSupplies(600);
			
		//Break an unwanted weapon down for supplies
		}else if (refCargo.getSupplies() < SUPL_LIMIT || Math.random() < 0.3)
			breakDown();
		
		//Build an wanted weapon with supplies
		else if (Math.random() < 0.5)
			buildWep();
		
		//Build either a ship or a fighter wing.
		else if (Math.random() < (float)makeShip.length/(makeShip.length+makeWing.length))
			shipMake(FleetMemberType.SHIP, makeShip[(int)(Math.random()*makeShip.length)]);
		else 
			shipMake(FleetMemberType.FIGHTER_WING, makeWing[(int)(Math.random()*makeWing.length)]);

		return null;
	}

	private Boolean isListed(String ID, String[] list){
		for(int i = 0; i < list.length; i++)
			if(ID.equals(list[i]))
				return true;
		return false;
	}
	
	private class ItemComp implements Comparator{
		public int compare(Object item1, Object item2){
			int item1num = (int)((CargoStackAPI)item1).getSize();
			int item2num = (int)((CargoStackAPI)item2).getSize();
			
			if(item1num > item2num)
				return 1;
			else if(item1num < item2num)
				return -1;
			else 
				return 0;			
		}
	}
	
	private void breakDown() {
		FleetDataAPI dockyard = refCargo.getMothballedShips();
		List moths = dockyard.getMembersListCopy();
		FleetMemberAPI breakShip = null;
		
		for (int i = 0; i < moths.size(); i++) {
			breakShip = (FleetMemberAPI)moths.get(i);
			if(! isListed(breakShip.getSpecId(),makeShip) && 
				! isListed(breakShip.getSpecId(),makeWing)){
				if(Global.getSettings().getBoolean("devMode") && NAG) 
					Global.getSector().addMessage("Eating Ship: " + breakShip.getSpecId(), Color.YELLOW);
				refCargo.addSupplies(shipSupplyCost(breakShip) * 0.9f);
				refCargo.removeFuel(breakShip.getFleetPointCost() * 10);
				dockyard.removeFleetMember(breakShip);
				return;
			}
		}
		
		List wepList = refCargo.getStacksCopy();
		Collections.sort(wepList, new ItemComp());

		String breakWep;
		for (int i = 0; i < wepList.size(); i++ ){
			if (!((CargoStackAPI)wepList.get(i)).isWeaponStack()) continue;

			breakWep = (String) ((CargoStackAPI)wepList.get(i)).getData();
			if (! isListed(breakWep,makeWep)){
				if(Global.getSettings().getBoolean("devMode") && NAG) 
					Global.getSector().addMessage("Eating Weapon: " + breakWep, Color.YELLOW);
				refCargo.addSupplies(40);
				refCargo.removeFuel(25);
				refCargo.removeWeapons(breakWep, 1);
				return;
			}
		}
	}

	private void buildWep() {
		refCargo.removeFuel(50f);
		refCargo.removeSupplies(50f);
		refCargo.addWeapons(makeWep[(int)(Math.random()*makeWep.length)], 1);
	}

	private void shipMake(FleetMemberType wingOrShip, String ID) {
		FleetMemberAPI ship = Global.getFactory().createFleetMember(wingOrShip, ID);
		float fuelCost = ship.getFleetPointCost() * 15;
		float supCost = shipSupplyCost(ship);
		if(fuelCost < refCargo.getFuel() && supCost < refCargo.getSupplies()){
			if(Global.getSettings().getBoolean("devMode") && NAG) 
				Global.getSector().addMessage("Building Ship: " + ID, Color.YELLOW);
			refCargo.removeFuel(fuelCost);
			refCargo.removeSupplies(supCost);
			refCargo.getMothballedShips().addFleetMember(ship);
		} else breakDown(); // grr. Need More. 
	}

	private float shipSupplyCost(FleetMemberAPI ship) {
		float buf = 0;
		if (ship.isFighterWing())	buf = 8;
		else if (ship.isFrigate())	buf = 16;
		else if (ship.isDestroyer())buf = 35;
		else if (ship.isCruiser())	buf = 90;
		else 								buf = 145;

		buf *= ship.getFleetPointCost();
		return buf;
	}
	
	private static String[] makeWep = { "gedune_plasma", "gedune_repeater", "gedune_flarepd",
			"annihilator", "lightag", "taclaser", "irpulse", "gedune_maelstrom", "gedune_scythe",
			"phasebeam", "hurricane", "gedune_hipdlaser"};
	private static String[] makeShip = { "gedune_byakuri_Hull",
			"gedune_kitsune_Hull", "gedune_mozok_Hull", "gedune_kyirus_Hull", "gedune_mozok_f_Hull", "gedune_mozok_t_Hull",
			"gedune_nanda_Hull", "gedune_tenzen_Hull", "gedune_bakoros_Hull" };
	private static String[] makeWing = { "gedune_chua_wing", "gedune_duri_wing" };
}

data/shipsystems/scripts/Gedune_BurnDriveStats.java
package data.shipsystems.scripts;

import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.plugins.ShipSystemStatsScript;

public class Gedune_BurnDriveStats implements ShipSystemStatsScript {

	public void apply(MutableShipStatsAPI stats, String id, State state, float effectLevel) {
		if (state == ShipSystemStatsScript.State.OUT) {
			stats.getMaxSpeed().unmodify(id); // to slow down ship to its regular top speed while powering drive down
		} else {
			stats.getMaxSpeed().modifyFlat(id, 240f * effectLevel);
			stats.getAcceleration().modifyFlat(id, 240f * effectLevel);
			//stats.getAcceleration().modifyPercent(id, 240f * effectLevel);
		}
	}
	public void unapply(MutableShipStatsAPI stats, String id) {
		stats.getMaxSpeed().unmodify(id);
		stats.getMaxTurnRate().unmodify(id);
		stats.getTurnAcceleration().unmodify(id);
		stats.getAcceleration().unmodify(id);
		stats.getDeceleration().unmodify(id);
	}
	
	public StatusData getStatusData(int index, State state, float effectLevel) {
		if (index == 0) {
			return new StatusData("increased engine power", false);
		}
		return null;
	}
}


data/missions/forgreatertesting/MissionDefinition.java
package data.missions.forgreatertesting;

import com.fs.starfarer.api.campaign.CargoAPI.CrewXPLevel;
import com.fs.starfarer.api.fleet.FleetGoal;
import com.fs.starfarer.api.fleet.FleetMemberType;
import com.fs.starfarer.api.mission.FleetSide;
import com.fs.starfarer.api.mission.MissionDefinitionAPI;
import com.fs.starfarer.api.mission.MissionDefinitionPlugin;

public class MissionDefinition implements MissionDefinitionPlugin {

	public void defineMission(MissionDefinitionAPI api) {

		// Set up the fleets so we can add ships and fighter wings to them.
		// In this scenario, the fleets are attacking each other, but
		// in other scenarios, a fleet may be defending or trying to escape
		api.initFleet(FleetSide.PLAYER, "ISS", FleetGoal.ATTACK, false, 5);
		api.initFleet(FleetSide.ENEMY, "ESS", FleetGoal.ATTACK, true, 3);

		// Set a small blurb for each fleet that shows up on the mission detail and
		// mission results screens to identify each side.
		api.setFleetTagline(FleetSide.PLAYER, "Testing Forces of the Republic");
		api.setFleetTagline(FleetSide.ENEMY, "Analysis Forces of the Empire");
		
		// These show up as items in the bulleted list under 
		// "Tactical Objectives" on the mission detail screen
		api.addBriefingItem("For Testing");
		
		// Set up the player's fleet.  Variant names come from the
		// files in data/variants and data/variants/fighters
		api.addToFleet(FleetSide.PLAYER, "gedune_nanda_variant1", FleetMemberType.SHIP, "ISS Black Star", true, CrewXPLevel.VETERAN);
		api.addToFleet(FleetSide.PLAYER, "gedune_duri_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "gedune_duri_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "gedune_duri_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "gedune_chua_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "gedune_chua_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "talon_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "talon_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.PLAYER, "talon_wing", FleetMemberType.FIGHTER_WING, false);
		
		// Mark both ships as essential - losing either one results
		// in mission failure. Could also be set on an enemy ship,
		// in which case destroying it would result in a win.
		api.defeatOnShipLoss("ISS Black Star");
		
		// Set up the enemy fleet.
		/*
      api.addToFleet(FleetSide.ENEMY, "wasp_wing", FleetMemberType.FIGHTER_WING, false);
      api.addToFleet(FleetSide.ENEMY, "wasp_wing", FleetMemberType.FIGHTER_WING, false);
      api.addToFleet(FleetSide.ENEMY, "gedune_chua_wing", FleetMemberType.FIGHTER_WING, false);
      api.addToFleet(FleetSide.ENEMY, "gedune_chua_wing", FleetMemberType.FIGHTER_WING, false);
		api.addToFleet(FleetSide.ENEMY, "talon_wing", FleetMemberType.FIGHTER_WING, false);*/
      api.addToFleet(FleetSide.ENEMY, "talon_wing", FleetMemberType.FIGHTER_WING, false);
      api.addToFleet(FleetSide.ENEMY, "talon_wing", FleetMemberType.FIGHTER_WING, false);
      api.addToFleet(FleetSide.ENEMY, "talon_wing", FleetMemberType.FIGHTER_WING, false);
		
		// Set up the map.
		float width = 6000f;
		float height = 3000f;
		api.initMap((float)-width/2f, (float)width/2f, (float)-height/2f, (float)height/2f);
		
		float minX = -width/2;
		float minY = -height/2;
		// And a few random ones to spice up the playing field.
		// A similar approach can be used to randomize everything
		// else, including fleet composition.
		for (int i = 0; i < 7; i++) {
			float x = (float) Math.random() * width - width/2;
			float y = (float) Math.random() * height - height/2;
			float radius = 100f + (float) Math.random() * 800f; 
			api.addNebula(x, y, radius);
		}
		
		// Add objectives. These can be captured by each side
		// and provide stat bonuses and extra command points to
		// bring in reinforcements.
		// Reinforcements only matter for large fleets - in this
		// case, assuming a 100 command point battle size,
		// both fleets will be able to deploy fully right away.
      
		api.addObjective(minX + width * 0.5f, minY + height * 0.5f, "nav_buoy");
		/*
      api.addObjective(minX + width * 0.75f, minY + height * 0.75f, "sensor_array");
		api.addObjective(minX + width * 0.4f, minY + height * 0.5f, "nav_buoy");
		api.addObjective(minX + width * 0.6f, minY + height * 0.5f, "nav_buoy");
		*/
		// Add an asteroid field going diagonally across the
		// battlefield, 2000 pixels wide, with a maximum of 
		// 100 asteroids in it.
		// 20-70 is the range of asteroid speeds.
		api.addAsteroidField(minX, minY + height * 0.5f, 0, 1000f,
								20f, 70f, 70);
		
		// Add some planets.  These are defined in data/config/planets.json.
		//api.addPlanet(minX + width * 0.2f, minY + height * 0.8f, 320f, "star_yellow", 300f);
		//api.addPlanet(minX + width * 0.8f, minY + height * 0.8f, 256f, "desert", 250f);
		//api.addPlanet(minX + width * 0.25f, minY + height * 0.5f, 200f, "barren", 200f);
	}

}


data/scripts/world/SectorGenWithGedune.java
package data.scripts.world;

import com.fs.starfarer.api.Global;
import java.awt.Color;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import com.fs.starfarer.api.FactoryAPI;
import com.fs.starfarer.api.campaign.*;
import com.fs.starfarer.api.fleet.*;

import data.scripts.world.*;

public class SectorGenWithGedune implements SectorGeneratorPlugin
{
	public void generate( SectorAPI sector )
	{
		StarSystemAPI system = sector.getStarSystem( "Corvus" );

		SectorEntityToken star = system.createToken( 0, 0 );
      
      SectorEntityToken planet = system.addPlanet( star, "Unidentified", "toxic", 270, 75, 9600, 150);
      SectorEntityToken GDstation = system.addOrbitalStation(planet,45, 300, 50,"Refactory", "gedune");

		GeduneSpawnPoint GDspawn = new GeduneSpawnPoint( sector, system, 6, 8, GDstation );
      
		system.addSpawnPoint( GDspawn );
      
		for( int i = 0; i < 5; i++ ) GDspawn.spawnFleet();
		
      CargoAPI GDcargo = GDstation.getCargo();
      GDcargo.addWeapons("gedune_plasma", (int)(Math.random()*4)+1);
      GDcargo.addWeapons("gedune_repeater", (int)(Math.random()*3)+3);
      GDcargo.addWeapons("gedune_flarepd", (int)(Math.random()*3)+3);
      GDcargo.addWeapons("annihilator", (int)(Math.random()*5)+5);
      GDcargo.addWeapons("lightag", (int)(Math.random()*3)+5);
      GDcargo.addWeapons("taclaser", (int)(Math.random()*3)+5);
      GDcargo.addWeapons("irpulse", (int)(Math.random()*3)+5);
      GDcargo.addSupplies((int)(Math.random()*150)+150);
      GDcargo.addMarines((int)(Math.random()*300)+400);
      GDcargo.addFuel((int)(Math.random()*1500)+1500);
      GDcargo.addCrew(CargoAPI.CrewXPLevel.GREEN,(int)(Math.random()*300)+200);
      GDcargo.addCrew(CargoAPI.CrewXPLevel.REGULAR,(int)(Math.random()*120)+60);
      GDcargo.addCrew(CargoAPI.CrewXPLevel.VETERAN,(int)(Math.random()*40)+20);
      
      //gedune steal stuff, we assume they've stolen some stuff already
      List weaponIds = sector.getAllWeaponIds();
		for (int i = 0; i < 7; i++) {
			String weaponId = (String) weaponIds.get((int) (weaponIds.size() * Math.random()));
			int quantity = (int)(Math.random() * 3f + 1f);
			GDcargo.addWeapons(weaponId, quantity);
		}
      
      FactoryAPI fac = Global.getFactory();
      FleetDataAPI GDmoths = GDcargo.getMothballedShips();
      String[] ships = new String[] {"gedune_chua_wing", "gedune_chua_wing", "gedune_duri_wing", "gedune_duri_wing", "gedune_duri_wing"};
      for(int i = 0; i < ships.length; i++)
         GDmoths.addFleetMember(fac.createFleetMember(FleetMemberType.FIGHTER_WING, ships[i]));
      
      ships = new String[]{"gedune_tenzen", "gedune_nanda", "gedune_nanda", "gedune_nanda", "gedune_mozok", "gedune_mozok_t", "gedune_mozok_f", "gedune_mozok_f", "gedune_kitsune", "gedune_kitsune"};
      for(int i = 0; i < ships.length; i++)
         GDmoths.addFleetMember(fac.createFleetMember(FleetMemberType.SHIP, ships[i] + "_Hull"));
         
		FactionAPI gedune = sector.getFaction( "gedune" );
      //Gedune national policy: "EVERYTHING MUST GO!"
      String[] peopleIhate = new String[] { "hegemony", "tritachyon", "pirates", "pirates", "independent", "player" };
      for(int i = 0; i < peopleIhate.length; i++)
         gedune.setRelationship( peopleIhate[i], -1); //I hate you

      //peopleIhate = new String[] { ... }; //This is the normal way to do it
      String[] peopleIhate2HATEHARDER = {"interstellarFederation", "junk_pirates", "lotus_pirates", "nomads", "serenitystation", "antediluvian", "ghost", "neutrinocorp", "syndicate_asp", "shadow_industry"}; //This is funnier.
      for (int i = 0; i < peopleIhate2HATEHARDER.length; i++)
         gedune.setRelationship( peopleIhate2HATEHARDER[i], -1);
      
      //incidentally, you can set any relationship you want, with anybody,
      //even people that don't exist. I can't believe that doesn't crash! 
      //Alex, Bro, THAT IS WHAT I CALL GRACEFUL FAILURE!
	}
	
}
